whd/dispatcher.py

```diff
@@ -318,6 +318,56 @@ def is_up_to_date(resource, resource_versions):
             sleep_seconds=sleep_seconds*1.2,
         )
 
+    def _jobs_not_triggered(self, concourse_api, matching_resources, pr_event):
+        '''
+        return all jobs and the corresponding resource versions, which have not been triggered
+        '''
+        for resource in matching_resources:
+            pipeline_name = resource.pipeline.name
+            resource_versions = concourse_api.resource_versions(
+                pipeline_name=pipeline_name,
+                resource_name=resource.name
+            )
+            pr_resource_versions = [
+                rv for rv in resource_versions if rv.version()['pr'] == str(pr_event.number())
+            ]
+            if not pr_resource_versions:
+                logger.warning(
+                    f"no resource versions found for pipeline {resource.pipeline.name} "
+                    f"resource {resource.name} and PR {pr_event.number()}"
+                )
+                return
+
+            # get all jobs which get triggered by the resource
+            jobs_triggered_by_resource = [
+                job for job in resource.pipeline.jobs if job.is_triggered_by_resource(resource.name)
+            ]
+
+            # check if there is a pr resource version which has not triggered a job.
+            # If so return the job and the resource version
+            for job in jobs_triggered_by_resource:
+                builds = concourse_api.job_builds(pipeline_name, job.name)
+                for resource_version in self._resource_versions_without_build(
+                    resource_versions=pr_resource_versions,
+                    builds=builds,
+                ):
+                    yield (job, resource, resource_version)
+
+    def _resource_versions_without_build(self, resource_versions, builds):
+        '''
+        yield resource versions for which no build exists by checking if the resource version
+        reference is found in any existing build
+        '''
+        for resource_version in resource_versions:
+            resource_version_in_build = False
+            for build in reversed(builds[-30:]):
+                build_plan = build.plan()
+                if build_plan.contains_version_ref(resource_version.version()['ref']):
+                    resource_version_in_build = True
+                    break
+            if not resource_version_in_build:
+                yield resource_version
+
```

Comment:
Haven't tested it, but think about something like
```python
def _jobs_not_triggered(self, resource_versions, builds):
    builds_to_consider = reversed(builds[-30:])

    yield from [
        rv for rv in resource_versions
        if not any([
            b.plan().contains_version_ref(rv.version()['ref']) for b in builds_to_consider
        ])
    ]
```
This would be a bit quicker to understand ,in my opinion.

Not a deal-breaker, though. I'll leave it up to you.